sklearn/metrics/_dist_metrics.pyx.tp

```diff
@@ -433,36 +721,36 @@ cdef class DistanceMetric{{name_suffix}}:
 
         Parameters
         ----------
-        X : array-like
-            Array of shape (Nx, D), representing Nx points in D dimensions.
-        Y : array-like (optional)
-            Array of shape (Ny, D), representing Ny points in D dimensions.
+        X : ndarray or CSR matrix of shape (n_samples_X, n_features)
+            Input data.
+        Y : ndarray or CSR matrix of shape (n_samples_Y, n_features)
+            Input data.
             If not specified, then Y=X.
 
         Returns
         -------
-        dist : ndarray
-            The shape (Nx, Ny) array of pairwise distances between points in
-            X and Y.
+        dist : ndarray of shape (n_samples_X, n_samples_Y)
+            The distance matrix of pairwise distances between points in X and Y.
         """
-        cdef cnp.ndarray[{{DTYPE_t}}, ndim=2, mode='c'] Xarr
-        cdef cnp.ndarray[{{DTYPE_t}}, ndim=2, mode='c'] Yarr
-        cdef cnp.ndarray[{{DTYPE_t}}, ndim=2, mode='c'] Darr
+        Y = X if Y is None else Y
+        X_is_Y = X is Y
+        X = check_array(X, accept_sparse=['csr'])
 
-        Xarr = np.asarray(X, dtype={{DTYPE}}, order='C')
-        self._validate_data(Xarr)
-        if Y is None:
-            Darr = np.zeros((Xarr.shape[0], Xarr.shape[0]),
-                         dtype={{DTYPE}}, order='C')
-            self.pdist(Xarr, Darr)
+        if X_is_Y:
+            Y = X
         else:
-            Yarr = np.asarray(Y, dtype={{DTYPE}}, order='C')
-            self._validate_data(Yarr)
-            Darr = np.zeros((Xarr.shape[0], Yarr.shape[0]),
-                         dtype={{DTYPE}}, order='C')
-            self.cdist(Xarr, Yarr, Darr)
-        return Darr
+            Y = check_array(Y, accept_sparse=['csr'])
```

Comment:
This could be simplified to:

```python
X = check_array(X, accept_sparse=['csr'])
if Y is None:
   Y = X
else:
   Y = check_array(Y, accept_sparse=['csr'])
```

Reply:
Sorry I must have been misleading, I don't think DistanceMetrics should do validation on X and Y. My thinking was:
DistanceMetrics expects sparse matrices in CSR format. It's the role of the estimator using DistanceMetrics (or more directly PairwiseDistancesReduction) to make sure to pass the array in the right format. So this estimator must set ``accept_sparse=["csr"]`` in its call to ``_validate_data``.

Reply:
`DistanceMetrics.pairwise` is technically never used by estimators directly: I think all estimators that rely on `DistanceMetrics` sublclasses either go through the `PairwiseDistanceReduction` class or the `BallTree` / `KDTree` classes and all of them call into the Cython API directly, without going through the public Python API.

However `DistanceMetrics.pairwise` is part of the public Python so I think it's fine to do input validation checks here.

Reply:
ah yes makes sense.

Can't wait for this to be deprecated :)